### PR TITLE
driver/telnet: Remove TELNET_CHARACTER_MODE

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary
If TELNET_CHARACTER_MODE is set, the ECHO must be handled by the telnet server.

So this option is useless. We need a new approach to control the ECHO behavior.

So a part of termios implement (local mode) to enable/disable ECHO.

## Impact
telnel driver only
## Testing
CI and custom boards

### CODE REVIEW

**Comment by pkarashchenko** [comment]
Do we need to create issue instead of simply adding comment?

**Comment by no1wudi** [comment]
> Do we need to create issue instead of simply adding comment?

We can create issue to track it, btw, I'll submit a fix this week.

File: drivers/net/telnet.c [hunk 1]
```diff
@@ -420,25 +415,16 @@ static ssize_t telnet_receive(FAR struct telnet_dev_s *priv,
             break;
 
           case STATE_DO:
-#ifdef CONFIG_TELNET_CHARACTER_MODE
-            if (ch == TELNET_SGA || ch == TELNET_ECHO)
+
+            if (priv->td_lflag & ECHO && ch == TELNET_ECHO)
```
Comment (pkarashchenko) [inline drivers/net/telnet.c:419]:
```suggestion
            if ((priv->td_lflag & ECHO) && ch == TELNET_ECHO)
```

**Comment by no1wudi** [comment]
@pkarashchenko @xiaoxiang781216 Could you review this PR again?

File: drivers/net/telnet.c [hunk 2]
```diff
@@ -1089,6 +1075,48 @@ static int telnet_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
...
+    case TCSETS:
+      {
+        termiosp = (FAR struct termios *)((uintptr_t)arg);
+        DEBUGASSERT(termiosp != NULL);
+
+        /* Save the termios settings */
+
+        priv->td_lflag = termiosp->c_lflag;
+
+        if (priv->td_lflag & ECHO)
```
Comment (pkarashchenko) [inline drivers/net/telnet.c:1100]: Optional
```suggestion
        if ((priv->td_lflag & ECHO) != 0)
```

**Review by pkarashchenko — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@no1wudi 

I noticed that an unnecessary ^J is shown when running telnet to esp32-devkitc:wifi_smp.
The symptom also happens with spresense:wifi_smp and spresense:rndis_smp.

```
Trying 192.168.10.30...
Connected to 192.168.10.30.
Escape character is '^]'.

NuttShell (NSH) NuttX-12.0.0
nsh> uname -a^JNuttX  12.0.0 c7a935dc1a Apr 24 2023 09:33:56 xtensa esp32-devkitc
nsh> cat /proc/uptime^J     46.30
nsh> ps^J  PID GROUP CPU PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK           STACK   USED  FILLED COMMAND
    0     0   0   0 FIFO     Kthread N-- Assigned           0000000000000000 002016 000768  38.0%  CPU0 IDLE
    1     1   1   0 FIFO     Kthread N-- Assigned           0000000000000000 002016 000432  21.4%  CPU1 IDLE
    2     2 --- 100 RR       Kthread --- Waiting  Mutex:3   0000000000000000 001984 000904  45.5%  lpwork 0x3ffb0570
    3     3   0 100 RR       Task    --- Running            0000000000000000 003008 002584  85.9%! nsh_main
    4     4 --- 223 RR       Kthread --- Waiting  Semaphore 0000000000000000 001984 000680  34.2%  rt_timer
    5     5 --- 253 RR       Kthread --- Waiting  MQ empty  0000000000000000 006600 001912  28.9%  wifi
    6     6 --- 100 RR       Task    --- Waiting  Semaphore 0000000000000000 001992 000952  47.7%  telnetd
   31    31 --- 100 RR       Task    --- Waiting  Semaphore 0000000000000000 001976 000808  40.8%  webserver
   43    43   1 100 RR       Task    --- Running            0000000000000000 003000 002120  70.6%  Telnet session
nsh> free^J                   total       used       free    largest  nused  nfree
        Umem:     240600      83296     157304     110000    198      8
nsh> ifconfig^Jwlan0	Link encap:Ethernet HWaddr ac:67:b2:71:65:2c at RUNNING
	inet addr:192.168.10.30 DRaddr:192.168.10.1 Mask:255.255.255.0

             IPv4   TCP   UDP  ICMP
Received     25ce  259d  0031  0000
Dropped      0000  0000  0000  0000
  IPv4        VHL: 0000   Frg: 0000
  Checksum   0000  0000  0000  ----
  TCP         ACK: 0000   SYN: 0000
              RST: 0000  0000
  Type       0000  ----  ----  0000
Sent         380a  3808  0002  0000
  Rexmit     ----  0005  ----  ----
nsh> exit^JConnection closed by foreign host.
```

**Comment by no1wudi** [comment]
@masayuki2009 
It seems that your telnet client works in character mode now instead of line mode, I'll have a test on esp32 to find why on these platform the telnet client works in different mode with  rv-virt:netnsh_smp and sim.

**Comment by no1wudi** [comment]
@masayuki2009 What's the version of your telnet client ? I'm using this version, and can't reproduce the issue on  esp32-devkitc:wifi_smp.
```
telnet (GNU inetutils) 2.4
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by many authors.
```

**Comment by masayuki2009** [comment]
@no1wudi 
I've been using the following package on ubuntu 20.04.6 LTS (amd64)
https://packages.ubuntu.com/focal/telnet

**Comment by no1wudi** [comment]
@masayuki2009 
I submit a fix here: https://github.com/apache/nuttx/pull/9133

Could you try it ?

**Pull request merged**